src/concepts/imperative.md

```diff
@@ -0,0 +1,167 @@
+`x` and `y` are both bound to `!pointer`, they have different values!
+
+> 💡 Comparison to the `void`!
+>
+> Remember that SML is a functional programming language. This means that
+> anything that's not an exception or infinite loop must be a value! As
+> such, when we do things with side effects (like assignment), we can't
+> really have _no value_ being returned. As such, SML implements this by
+> having these side effect functions and operations return `() : unit`.
+> Theoretically, `:=` could return something crazy, like an `'a tree list option`
+> if we really wanted to. It's just that `() : unit` is convenient.
+> So when you see `:= : 'a * 'a ref -> unit`, you can think of it as a
+> `void` method or function in Java, C, and C++.
+>
+> In other words, returning `() : unit` for the assignment operator is
+> SML's way of implementing `void` methods.
+
+Introducing imperative logic causes difficulties like these. It takes away the
+confidence we have in our values instilled in us by referential transparency.
+It often makes doing formal proofs about imperative code a bit more difficult.
+
+### Sequential Evaluation `;`
+
+Finally, having a way to execute side effects but still return something
+meaningful will be useful to use. Generally, `;` (yes this is just a
+semicolon lol) lets us "execute" two "programs" in sequential order.
+You see this used in the REPL:
+
+```
+val x = 2 + 2;
+val y = 8 - 4;
+```
+
+Where we execute one "program" with the side effect of binding `2 + 2`
+to `x`, and then execute the next "program" with the side effect of
+binding `8 - 4` to `y`, and continue to wait for the next "program".
+If you're familiar with languages like Java and C, this can also be
+used to explain why semicolons are required at the end of each line.
+
+Where the semicolon `;` is used to mark the end of one program and set up
+the next program to sequentially follow. In SML, this is also valid syntax.
+
+> For some expressions `exp1 : t1`, `exp2 : t2`, we have that `exp1;exp2 : t2`
+> where `exp1` is evaluated first, and then `exp2` evaluated thereafter.
+
+One thing to note is that in SML, only the value of the second expression
+is returned, and the type of the whole expression is the type of the last
+expression.
+
+```sml
+val pointer : int ref = ref 0
+val x : int = (pointer := 7; !pointer)
+val y : int = (pointer := 21; !pointer)
+```
+
+Here, we first initialize `pointer` to a dummy value. After that, we run
+the program `pointer := 7`, and then right after, return `!pointer`. This
+would mean that the value `7 : int` is bound to `x`, just like in the above
+examples. Similarly, we run The program `pointer := 21`, and then right
+after, return `!pointer`. This would mean that the value `21 : int` is
+bound to `y`, just like in the above examples.
+
+## Putting it All Together
+
+Here's an example that uses all of the above operations in a meaningful way.
+
+```sml
+local
```

Comment:
maybe worth noting that this implementation of fact has something wrong with it (it only works the first time it's called)